### PR TITLE
Parameter values get evaluated by yaml rules

### PR DESCRIPTION
**Pull request by sloretz**
Resolves ros2/launch_ros#30

This PR evaluates parameter values as `yaml` would.

I'm not set up to run mypy at the moment, so I'm not confident the type annotations are correct.

### CODE REVIEW

File: launch_ros/launch_ros/utilities/evaluate_parameters.py [hunk 1]
```diff
@@ -52,12 +54,29 @@ def evaluate_parameter_dict(
...
                 # Value is an array of a list of substitutions
                 output_subvalue = []  # List[str]
                 for subvalue in value:
-                    output_subvalue.append(perform_substitutions(context, list(subvalue)))
+                    value = perform_substitutions(context, list(subvalue))
+                    output_subvalue.append(value)
                     evaluated_value = tuple(output_subvalue)
+                # All values in a list must have the same type.
+                # If they don't then assume it is a list of strings
+                yaml_evaluated_value = []  # Union[List[str], List[int], List[float], List[bool]]
```
Comment (wjwwood) [inline launch_ros/launch_ros/utilities/evaluate_parameters.py]: nitpick: now that we don't have to support python3.5 we can use the normal syntax for this:

```python
yaml_evaluated_value: Union[List[str], List[int], List[float], List[bool]] = []
```

File: launch_ros/launch_ros/utilities/evaluate_parameters.py [hunk 2]
```diff
@@ -52,12 +54,29 @@ def evaluate_parameter_dict(
...
+                type_ = None
+                dissimilar_types = False
+                for subvalue in evaluated_value:
+                    yaml_subvalue = yaml.safe_load(subvalue)
+                    subtype = type(yaml_subvalue)
+                    if type_ is not None and type_ != subtype:
```
Comment (wjwwood) [inline launch_ros/launch_ros/utilities/evaluate_parameters.py]: When is `type_` `None`?

File: launch_ros/launch_ros/utilities/evaluate_parameters.py [hunk 3]
```diff
@@ -52,12 +54,29 @@ def evaluate_parameter_dict(
...
+                        dissimilar_types = True
+                        break
+                    yaml_evaluated_value.append(yaml_subvalue)
+                    type_ = subtype
+                if not dissimilar_types:
```
Comment (wjwwood) [inline launch_ros/launch_ros/utilities/evaluate_parameters.py]: What should happen (if anything) in the `else` case for if statement?

File: launch_ros/launch_ros/utilities/evaluate_parameters.py (hunk 1 above)
Comment (sloretz) [reply launch_ros/launch_ros/utilities/evaluate_parameters.py]: Replaced comments with new syntax in this function in 379e43f

File: launch_ros/launch_ros/utilities/evaluate_parameters.py (hunk 2 above)
Comment (sloretz) [reply launch_ros/launch_ros/utilities/evaluate_parameters.py]: `type_` is `None` when looking at the first element of the list. A better name would be `last_subtype`

Comment (sloretz) [reply launch_ros/launch_ros/utilities/evaluate_parameters.py]: `type_` -> `last_subtype` in 032ae59

File: launch_ros/launch_ros/utilities/evaluate_parameters.py (hunk 3 above)
Comment (sloretz) [reply launch_ros/launch_ros/utilities/evaluate_parameters.py]: Nothing should happen. If the values in a sequence after yaml'ing are not all the same type then `evaluated_value` stays a sequence of strings.

**Comment by sloretz** [comment]
CI (testing launch_ros and test_launch_ros)
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/7184/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/3395/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/5891/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/7021/)

**Comment by Karsten1987** [comment]
Oh, too late for the party. But anyway, this works great! :)

One can now set boolean parameters when launching a node:
```
        Node(
            package='robot_state_publisher',
            node_executable='robot_state_publisher',
            node_name='robot_state_publisher',
            output='screen',
            parameters=[{'use_sim_time': LaunchConfiguration('use_sim_time')}],
            arguments=[urdf])
```

**Pull request merged**